File: website/verify-build.py
#!/usr/bin/env python3
"""Check the combined Pages artifact without credentials or network access."""
import argparse
import json
from pathlib import Path, PurePosixPath


def verify(root):
    demo = root / "demo"
    if not (root / "index.html").is_file() or not (root / "sw.js").is_file():
        raise ValueError("Regular application/PWA build is missing")
    if not (demo / "index.html").is_file():
        raise ValueError("Demo build is missing")
    if any(path.is_file() and (path.name in {"sw.js", "manifest.webmanifest"} or path.name.startswith("workbox-")) for path in demo.rglob("*")):
        raise ValueError("Demo must not contain PWA artifacts")
    html = (demo / "index.html").read_text()
    if 'rel="manifest"' in html:
        raise ValueError("Demo HTML must not declare a PWA manifest")
    catalog = demo / "catalog"
    for required in ["search3.json", "assets.json", "credits.html", "credits.json", "dates.json", "sources.json"]:
        if not (catalog / required).is_file():
            raise ValueError("Demo catalog metadata is incomplete")
    assets = json.loads((catalog / "assets.json").read_text())
    for asset in assets.values():
        path = PurePosixPath(asset["path"])
        if path.is_absolute() or ".." in path.parts or path.parts[0] not in {"audio", "covers"}:
            raise ValueError("Unsafe demo asset path")
        file = catalog / path
        if not file.is_file() or file.is_symlink():
            raise ValueError("Missing demo audio or artwork")
    search = json.loads((catalog / "search3.json").read_text())["subsonic-response"]["searchResult3"]
    if not search["song"] or any(song["id"] not in assets for song in search["song"]):
        raise ValueError("Missing demo tracks")
    size = sum(path.stat().st_size for path in root.rglob("*") if path.is_file())
    if size > 1024**3:
        raise ValueError("Combined Pages site exceeds 1 GiB")
    return len(search["song"]), size


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--root", type=Path, default=Path("dist"))
    args = parser.parse_args()
    tracks, size = verify(args.root)
    print(f"Pages artifact verified: {tracks} demo tracks; {size / 1024**2:.1f} MiB total")

File: website/verify-build.mjs
import { lstat, readdir, readFile } from "node:fs/promises";
import { join, resolve } from "node:path";
import { parseArgs } from "node:util";

class ArtifactError extends Error {}

async function verify(root) {
  const files = new Set();
  let size = 0;
  async function walk(directory, prefix = "") {
    for (const name of await readdir(directory)) {
      const path = prefix + name;
      const info = await lstat(join(directory, name));
      if (info.isSymbolicLink())
        throw new ArtifactError("Pages artifact must not contain symlinks.");
      if (info.isDirectory()) await walk(join(directory, name), path + "/");
      else if (info.isFile()) {
        files.add(path);
        size += info.size;
      } else throw new ArtifactError("Unexpected Pages artifact entry.");
    }
  }
  await walk(root);
  for (const path of ["index.html", "webapp/index.html", "webapp/sw.js", "webapp/manifest.webmanifest"]) {
    if (!files.has(path))
      throw new ArtifactError("Regular application/PWA or demo build is missing.");
  }
  for (const path of files) {
    const name = path.split("/").at(-1);
    if (
      !path.startsWith("webapp/") &&
      (name === "sw.js" || name === "manifest.webmanifest" || name.startsWith("workbox-"))
    ) {
      throw new ArtifactError("Demo must not contain PWA artifacts.");
    }
  }
  const read = (path) => readFile(join(root, path), "utf8");
  if ((await read("index.html")).includes('rel="manifest"')) {
    throw new ArtifactError("Demo HTML must not declare a PWA manifest.");
  }
  const manifest = JSON.parse(await read("webapp/manifest.webmanifest"));
  if ([manifest.id, manifest.start_url, manifest.scope].some((path) => path !== "/webapp/")) {
    throw new ArtifactError("PWA must launch and stay scoped to /webapp/.");
  }
  if (!(await read("webapp/index.html")).includes('href="/webapp/manifest.webmanifest"')) {
    throw new ArtifactError("App HTML must declare its scoped PWA manifest.");
  }
  for (const name of [
    "search3.json",
    "assets.json",
    "credits.html",
    "credits.json",
    "dates.json",
    "sources.json",
  ]) {
    if (!files.has("catalog/" + name))
      throw new ArtifactError("Demo catalog metadata is incomplete.");
  }
  if (size > 1024 ** 3) throw new ArtifactError("Combined Pages site exceeds 1 GiB.");
  console.log(`Pages artifact verified: ${(size / 1024 ** 2).toFixed(1)} MiB total`);
}

try {
  const { values } = parseArgs({ options: { root: { type: "string", default: "dist" } } });
  await verify(resolve(values.root));
} catch (error) {
  console.error(
    "Pages artifact verification failed:",
    error instanceof ArtifactError ? error.message : "Missing, unsafe or invalid build output.",
  );
  process.exitCode = 1;
}
